src/kernel.ts

```diff
@@ -1,1069 +1,803 @@
-         */
-        this.id = data.id;
-        this.kernel_url = utils.url_join_encode(this.kernel_service_url, this.id);
-        this.start_channels();
-    }
 
-    _kernel_connected(): void {
-        /**
-         * Perform necessary tasks once the connection to the kernel has
-         * been established. This includes requesting information about
-         * the kernel.
-         *
-         * @function _kernel_connected
-         */
-        events.trigger('kernel_connected.Kernel', {kernel: this});
-        // get kernel info so we know what state the kernel is in
-        var that = this;
-        this.kernel_info((reply?) => {
-            this.info_reply = reply.content;
-            events.trigger('kernel_ready.Kernel', {kernel: that});
+/**
+ * A Kernel class to communicate with the Python kernel. This
+ * should generally not be constructed directly, but be created
+ * by the `Session` object. Once created, this object should be
+ * used to communicate with the kernel.
+ */
+export
+class Kernel {
+
+  static statusChange = new Signal<Kernel, string>();
+
+  /**
+   * GET /api/kernels
+   *
+   * Get the list of running kernels.
+   */
+  static list(kernelServiceUrl: string): Promise<any> {
+    return utils.ajaxRequest(kernelServiceUrl, {
+      method: "GET",
+      dataType: "json"
+    });
+  }
+
+  constructor(kernelServiceUrl: string, wsUrl: string, name: string) {
+    this._id = null;
+    this._name = name;
+    this._ws = null;
+
+    this._kernelServiceUrl = kernelServiceUrl;
+    this._kernelUrl = null;
+    this._wsUrl = wsUrl;
+    if (!this._wsUrl) {
+      // trailing 's' in https will become wss for secure web sockets
+      this._wsUrl = location.protocol.replace('http', 'ws') + "//" + location.host;
+    }
+
+    this._username = "username";
+    this._sessionId = utils.uuid();
+    this._infoReply = {}; // kernel_info_reply stored here after starting
+    this._handlerMap = new Map<string, KernelFutureHandler>();
+
+    if (typeof WebSocket === 'undefined') {
+      alert('Your browser does not have WebSocket support, please try Chrome, Safari, or Firefox ≥ 11.');
+    }
+
+    this._autorestartAttempt = 0;
+    this._reconnectAttempt = 0;
+    this._reconnectLimit = 7;
+  }
+
+  /**
+   * Get the Url for the Kernel service.
+   */
+  get kernelServiceUrl() : string {
+    return this._kernelServiceUrl;
+  }
+
+  /**
+   * Check whether there is a connection to the kernel. This
+   * function only returns true if websocket has been
+   * created and has a state of WebSocket.OPEN.
+   */
+  get isConnected(): boolean {
+    // if any channel is not ready, then we're not connected
+    if (this._ws === null) {
+      return false;
+    }
+    if (this._ws.readyState !== WebSocket.OPEN) {
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Check whether the connection to the kernel has been completely
+   * severed. This function only returns true if all channel objects
```

Comment:
"all channel objects are" -> "the websocket is".  Again, this is leftover from when we had multiple websockets.